src/auth/src/credentials/user_credential.rs

```diff
@@ -78,24 +78,24 @@ impl TokenProvider for UserTokenProvider {
             .request(Method::POST, self.endpoint.as_str())
             .header(CONTENT_TYPE, header)
             .json(&req);
-        let resp = builder
-            .send()
-            .await
-            .map_err(CredentialError::non_retryable)?;
+        let resp = builder.send().await.map_err(CredentialError::retryable)?;
 
         // Process the response
         if !resp.status().is_success() {
             let status = resp.status();
-            let body = resp.text().await.map_err(CredentialError::non_retryable)?;
+            let body = resp
+                .text()
+                .await
+                .map_err(|e| CredentialError::new(is_retryable(status), e.into()))?;
             return Err(CredentialError::new(
                 is_retryable(status),
                 Box::from(format!("Failed to fetch token. {body}")),
             ));
         }
-        let response = resp
-            .json::<Oauth2RefreshResponse>()
-            .await
-            .map_err(CredentialError::non_retryable)?;
+        let response = resp.json::<Oauth2RefreshResponse>().await.map_err(|e| {
+            let retryable = !e.is_decode();
+            CredentialError::new(retryable, e.into())
+        })?;
```

Comment:
I am not sure if this helps, but you could implement the `From` trait:

```rust
impl std::convert::From<reqwest::Error> for CredentialError {
  fn from(value: reqwest::Error) -> Self {
    // Do the needful
  }
}
```


And then you can write (I think) `.map_err(CredentialError::from)`.  The `from` function is also easier to test.  But whether it works in all these cases? 🤷